types/react/experimental.d.ts

```diff
@@ -38,7 +38,24 @@ import React = require('.');
 
 export {};
 
+type NonNullableValue = string | number | boolean | object | symbol;
+
 declare module '.' {
+    type MutableSource<TSource extends NonNullableValue, TVersion extends NonNullableValue> = {
+        _source: TSource,
```

Comment:
Since they are private I would like to omit them and find another way to "tag" or "brand" a MutableSource. I would rather add a non-existent property than type private properties.

Reply:
@eps1lon  How about this

```ts
    type MutableSource<Source extends NonNullableValue, Version extends NonNullableValue> = Omit<{
        _source: Source,
        _getVersion: (source: Source) => Version,
    }, '_source' | '_getVersion'>;
```

Now source doesn't allow access to these fields but TypeScript still infers templates as well.

Reply:
Equals to:

```ts
type MutableSource<Source extends NonNullableValue, Version extends NonNullableValue> = Pick<{
        _source: Source,
        _getVersion: (source: Source) => Version,
    }, never>;
```

Reply:
But then it would allow any object shape, right? I don't think it's worth spending time on this issue. As long as we don't have type "branding" or "tagging" we can keep this approach. I would still prefer a JSDOC `@private` annotation though. That might even help since it could be hidden from IntelliSense.